syft/frameworks/torch/pointers/pointer_tensor.py

```diff
@@ -0,0 +1,181 @@
+
+    A PointerTensor forwards all API calls to the remote.PointerTensor objects
+    point to tensors (as their name implies). They exist to mimic the entire
+    API of a normal tensor, but instead of computing a tensor function locally
+    (such as addition, subtraction, etc.) they forward the computation to a
+    remote machine as specified by self.location. Specifically, every
+    PointerTensor has a tensor located somewhere that it points to (they should
+    never exist by themselves). Note that PointerTensor objects can point to
+    both torch.Tensor objects AND to other PointerTensor objects. Furthermore,
+    the objects being pointed to can be on the same machine or (more commonly)
+    on a different one. Note further that a PointerTensor does not know the
+    nature how it sends messages to the tensor it points to (whether over
+    socket, http, or some other protocol) as that functionality is abstracted
+    in the BaseWorker object in self.location.
+
+    Example:
+
+     >>> import syft as sy
+     >>> hook = sy.TorchHook()
+     >>> bob = sy.VirtualWorker(id="bob")
+     >>> x = sy.Tensor([1,2,3,4,5])
+     >>> y = sy.Tensor([1,1,1,1,1])
+     >>> x_ptr = x.send(bob) # returns a PointerTensor, sends tensor to Bob
+     >>> y_ptr = y.send(bob) # returns a PointerTensor, sends tensor to Bob
+     >>> # executes command on Bob's machine
+     >>> z_ptr = x_ptr + y_ptr
+    """
+
+    def __init__(
+        self,
+        location: "AbstractWorker" = None,
+        id_at_location: Union[str, int] = None,
+        owner: "AbstractWorker" = None,
+        id: Union[str, int] = None,
+        garbage_collect_data: bool = True,
+        shape: torch.Size = None,
+        point_to_attr: str = None,
+        tags: List[str] = None,
+        description: str = None,
+    ):
+        super().__init__(
+            location=location,
+            id_at_location=id_at_location,
+            owner=owner,
+            id=id,
+            garbage_collect_data=garbage_collect_data,
+            point_to_attr=point_to_attr,
+            tags=tags,
+            description=description,
+        )
+        self._shape = shape
+
+    def get_shape(self):
+        """Request information about the shape to the remote worker"""
+        return self.owner.request_remote_tensor_shape(self)
+
+    @property
+    def shape(self):
+        """ This method returns the shape of the data being pointed to.
+        This shape information SHOULD be cached on self._shape, but
+        occasionally this information may not be present. If this is the
+        case, then it requests the shape information from the remote object
+        directly (which is inefficient and should be avoided).
+        """
+
+        if self._shape is None:
+            self._shape = self.get_shape()
+
+        return self._shape
+
+    @shape.setter
+    def shape(self, new_shape):
+        self._shape = new_shape
+
+    def get(self, deregister_ptr: bool = True):
```

Comment:
how does this differ in logic from the one in ObjectPointer

Reply:
I refactored the code so that the difference is clearly visible.